In [ ]:
import pandas as pd
import numpy as np
import matplotlib as mp
import statsmodels.api as sm

from scipy.linalg import toeplitz
from statsmodels.tsa.stattools import acf

In [ ]:
df = pd.read_csv('https://raw.githubusercontent.com/xiaowanzisj/Data-Science-1/refs/heads/main/dataset_lm.csv')

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 422 entries, 0 to 421
Data columns (total 16 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   Dependent Var        422 non-null    float64
 1   Explanatory Var #1   422 non-null    float64
 2   Explanatory Var #2   422 non-null    float64
 3   Explanatory Var #3   422 non-null    int64  
 4   Explanatory Var #4   422 non-null    float64
 5   Explanatory Var #5   422 non-null    float64
 6   Explanatory Var #6   422 non-null    float64
 7   Explanatory Var #7   422 non-null    float64
 8   Explanatory Var #8   422 non-null    int64  
 9   Explanatory Var #9   422 non-null    float64
 10  Explanatory Var #10  422 non-null    float64
 11  Explanatory Var #11  422 non-null    float64
 12  Explanatory Var #12  422 non-null    float64
 13  Explanatory Var #13  422 non-null    int64  
 14  Explanatory Var #14  422 non-null    float64
 15  Explanatory Var #15  422 non-null    flo

In [ ]:
df

Dependent Var  Explanatory Var #1  Explanatory Var #2  \
0        56.293458           13.698667           50.639873   
1        58.473431            2.714725           65.845845   
2        94.195330           11.618072           65.072497   
3        29.074583            0.818623           45.408996   
4        86.035569            9.077544           73.548021   
..             ...                 ...                 ...   
417      61.300432            0.338441           70.430598   
418      26.309237           -1.729712           47.087996   
419      58.350627           18.322301           53.267835   
420      31.954003            0.436357           61.844132   
421      50.590958           -3.107425           64.590632   

     Explanatory Var #3  Explanatory Var #4  Explanatory Var #5  \
0                     0          -18.568035           45.121911   
1                     1          -25.105932           47.190213   
2                     0           -7.897464           52.163036   
3                     1          -18.316132           54.356714   
4                     0          -19.204165           47.186807   
..                  ...                 ...                 ...   
417                   0          -21.525979           50.741115   
418                   1          -19.034807           55.242929   
419                   0          -26.186201           36.702958   
420                   0          -25.881961           57.106851   
421                   1          -20.303215           56.374124   

     Explanatory Var #6  Explanatory Var #7  Explanatory Var #8  \
0             11.412501           56.410757                   2   
1             10.080280           65.383107                   3   
2             11.057301           82.812717                   0   
3              5.029029           48.812471                   1   
4             12.128134           62.520911                   2   
..                  ...                 ...                 ...   
417           -8.050843           39.075397                   3   
418           28.001769           76.429649                   0   
419           14.530116           51.275150                   2   
420           21.786066           68.928442                   0   
421            4.604237           49.501005                   0   

     Explanatory Var #9  Explanatory Var #10  Explanatory Var #11  \
0            -12.281132            38.996909            -3.010548   
1            -36.763585            51.654939             4.991111   
2            -15.733547            48.913837            -2.457696   
3            -12.825591            45.851732            14.974177   
4            -13.804860            47.765904             9.593982   
..                  ...                  ...                  ...   
417          -29.197173            57.473862            15.505202   
418          -21.941401            54.641620            14.295688   
419          -20.699670            46.443687             5.708762   
420          -23.166461            59.194840             8.365762   
421          -18.792745            45.594612            16.414449   

     Explanatory Var #12  Explanatory Var #13  Explanatory Var #14  \
0              49.195073                    0           -21.153143   
1              45.591729                    0            -6.474403   
2              56.608806                    0           -27.903299   
3              47.362594                    1           -10.064411   
4              53.700562                    0           -17.546302   
..                   ...                  ...                  ...   
417            73.280605                    1            -3.602482   
418            49.816798                    1           -13.910494   
419            52.751016                    1            -6.202345   
420            64.291056                    1           -21.330017   
421            66.957547                    0      

In [ ]:
df.isna().sum()

Dependent Var          0
Explanatory Var #1     0
Explanatory Var #2     0
Explanatory Var #3     0
Explanatory Var #4     0
Explanatory Var #5     0
Explanatory Var #6     0
Explanatory Var #7     0
Explanatory Var #8     0
Explanatory Var #9     0
Explanatory Var #10    0
Explanatory Var #11    0
Explanatory Var #12    0
Explanatory Var #13    0
Explanatory Var #14    0
Explanatory Var #15    0
dtype: int64

### Part A (25%)
1.	Download the “dataset_lm.csv” file from Canvas and upload it to Jupyter Notebook. 
2.	Run the OLS model by using the dependent and explanatory variables in the dataset. 
3.	Show your summary table in Python and interpret your results in the summary report.


In [ ]:
y = df['Dependent Var']
x = df.drop('Dependent Var', axis=1)

In [ ]:
x_updated = sm.add_constant(x)
model_updated = sm.OLS(y,x_updated).fit()

In [ ]:
print(model_updated.summary())

                            OLS Regression Results                            
Dep. Variable:          Dependent Var   R-squared:                       1.000
Model:                            OLS   Adj. R-squared:                  1.000
Method:                 Least Squares   F-statistic:                 1.682e+30
Date:                Fri, 18 Oct 2024   Prob (F-statistic):               0.00
Time:                        23:30:36   Log-Likelihood:                 12021.
No. Observations:                 422   AIC:                        -2.401e+04
Df Residuals:                     406   BIC:                        -2.395e+04
Df Model:                          15                                         
Covariance Type:            nonrobust                                         
                          coef    std err          t      P>|t|      [0.025      0.975]
---------------------------------------------------------------------------------------
const                  32.0000   9

1. R-squared (1.000) and Adjusted R-squared (1.000):
The R-squared value in a regression model measures how well the independent variables explain the variability in the dependent variable. A value of 1.000 (or 100%) means that explanatory variables perfectly predict the dependent variable, which means that the model captures all possible variations in the dependent variable and leaves no room for error or randomness. The model maybe overfitting the data. 
The Adjusted R-squared is 1.000, which indicates that the model perfectly explains the variability of the dependent variable, even after adjusting for the number of predictors.

2. F-statistic (1.682e+30) and Prob (F-statistic) (0.000):
The F-statistic tests whether the independent variables, as a group, significantly predict the dependent variable. A very large F-statistic with a Prob (F-statistic) of 0.000 indicates that the model is statistically significant, meaning the explanatory variables jointly have a strong influence on the dependent variable.

3. Coefficients (coef):
•	Significant coefficients include:
o	Explanatory Var #3: the most significant positive coefficient of 6.200, which means a 1-unit increase in this explanatory variable leads to an increase of 6.200 in the dependent variable, assuming all other variables are held constant.
o	Explanatory Var #5: the most significant negative coefficient of -0.900, indicating a negative relationship, where an increase in this variable decreases the dependent variable by 0.900.
These variables should be considered carefully when making decisions based on the model, as they are the most influential in either direction.
4. P-values (P>|t|):
•	Variables with p-values less than 0.05: If the p-value is less than 0.05, the variable is considered statistically significant, which means that the variable has a meaningful impact on the dependent variable. Explanatory variable #1,2,3,4,5,6,10,14,15.
•	Variables with p-values greater than 0.05 (like Explanatory Var #7, 8, 9, 11,12,13) are not statistically significant and do not have a strong impact on the dependent variable.

5. Standard Errors:
•	The standard error (std err) indicates how much variation exists in the coefficient estimate. Smaller standard errors suggest higher precision and confidence in the estimate of that variable.
•	Smaller values indicate more precise estimates of the corresponding coefficients, whereas larger values suggest more variability in the estimate. Variables like #3, #8, and #13 have larger standard errors, implying more uncertainty in their coefficient estimates, while variables like #1, #2, and #6 have very small standard errors, indicating higher precision.
6. T-statistics:
•  T-statistics help determine if a variable is important in predicting the outcome. The bigger the t-value (positive or negative), the more significant that variable is. If the t-value is small or close to 0, the variable probably doesn't have much impact on the dependent variable. Based on the OLS regression results, Explanatory Var #1 is highly significant with a t-value of 1.75e+15, indicating a strong influence on the dependent variable, while Explanatory Var #7 shows a t-value of 0.867, suggesting it has little to no significant impact on the model's prediction.


### Part B (25%)
1.	Use error values from the OLS model to calculate their standard deviation and autocorrelation values for the first three lags. 
2.	Then, run the GLS model accordingly. 
3.	Show your summary table in Python and interpret your results in the summary report.


In [ ]:
# Q1
residuals = model_updated.resid
std_deviation = residuals.std()
std_deviation

np.float64(6.591964783129596e-14)

In [ ]:
acf_residuals = acf(residuals, nlags = 3)
acf_residuals

array([ 1.        ,  0.04566151, -0.02558894, -0.0018177 ])

In [ ]:
# Q2
rho = acf_residuals[1]
n = len(y)
cov_matrix = std_deviation**2 * toeplitz(np.append([1, rho], np.zeros(n-2)))
print(sm.GLS(y, x_updated, sigma=cov_matrix).fit().summary())

                            GLS Regression Results                            
Dep. Variable:          Dependent Var   R-squared:                       1.000
Model:                            GLS   Adj. R-squared:                  1.000
Method:                 Least Squares   F-statistic:                 5.089e+29
Date:                Fri, 18 Oct 2024   Prob (F-statistic):               0.00
Time:                        23:30:36   Log-Likelihood:                 11769.
No. Observations:                 422   AIC:                        -2.351e+04
Df Residuals:                     406   BIC:                        -2.344e+04
Df Model:                          15                                         
Covariance Type:            nonrobust                                         
                          coef    std err          t      P>|t|      [0.025      0.975]
---------------------------------------------------------------------------------------
const                  32.0000   1

1. R-squared (1.000) & Adjusted R-squared (1.000):
Both R-squared and Adjusted R-squared are 1.000, which means that the model explains 100% of the variation in the dependent variable. This is highly unusual and often indicates overfitting, where the model is too perfectly fitted to the data and may not generalize well to new data.

2. F-statistic (5.089e+29) & Prob (F-statistic) (0.00):
F-statistic (5.089e+29): This is an extremely high value, meaning the variables in your model are likely doing a good job predicting the outcome. Probability (F-statistic) (0.00): The probability of this F-statistic happening by chance is essentially zero. This means the model is highly significant, and the variables together are important in explaining the outcome.

3. Coefficients (coef):
The GLS regression model shows that several explanatory variables, such as Var #1, #2, #3, and #4, have significant positive coefficients, indicating a strong positive correlation with the dependent variable. In contrast, Var #5 and Var #12 have significant negative coefficients, suggesting a negative correlation with the dependent variable. Other explanatory variables, such as Var #6, #7, #8, #9, and #10, are not statistically significant, implying they have little influence on the model. Overall, the R-squared value of 1.000 indicates a very high fit, but it may also suggest the risk of overfitting

4. P-values (P>|t|)：
The p-values help in identifying which variables have a statistically significant effect on the outcome of the model. Explanatory Var #1, #2, #3, #4, #5, and #12 all have p-values of 0.000 or close to it, indicating that these variables are statistically significant at the 0.05 level. Explanatory Var #6, #7, #8, #9, #10, #11, #13, #14, and #15 have higher p-values (greater than 0.05), indicating they are not statistically significant in this model

5. Standard Errors:
The standard errors in GLS regression output are very small, indicating high precision in estimating the coefficients. Each variable's standard error shows little variability, confirming the accuracy of the model's estimates.

6. T-statistics:
The large t-statistics for most coefficients, combined with their very low p-values, suggest that these variables are statistically significant and play an important role in explaining the dependent variable. 

### Part C (25%)
1.	Split the dataset into two as the training and test sets (test size = 0.5). 
2.	Run the Lasso model with alpha=1 and estimate the coefficients using the training set. 
3.	Then, calculate the mean absolute percentage error using the test set. 
4.	Find an approximate value for alpha that minimizes the mean absolute percentage error.

In [ ]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import Lasso
from sklearn.metrics import mean_absolute_percentage_error

In [ ]:
# Split the dataset into two as the training and test sets
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.5, random_state=42)

In [ ]:
lasso = Lasso(alpha=1)
lasso.fit(x_train, y_train)
lasso.coef_

array([ 1.26972628,  1.68394638,  2.02626245,  2.08756512, -0.91746375,
       -0.        ,  0.        , -0.        ,  0.        , -0.        ,
        0.01314162,  0.        , -0.        ,  0.        , -0.03617731])

Lasso has set some coefficients to zero, indicating that these variables may not contribute significantly to the model. 

In [ ]:
y_pred = lasso.predict(x_test)
mape = mean_absolute_percentage_error(y_test, y_pred)
mape

np.float64(0.04432190198291579)

1. The Mean Absolute Percentage Error (MAPE), which indicates the average percentage error between the predicted and actual values in the test set for the Lasso model on the test set, is approximately 4.43%. 
2. A 4.43% MAPE number seems reasonable, only depending on the context and the specific field or industry where the data comes from. 

In [ ]:
# Try alpha values between 0.001 and 100
alphas = np.logspace(-3, 2, 100)  
best_alpha = 1
best_mape = float('inf')

for alpha in alphas:
    lasso = Lasso(alpha=alpha)
    lasso.fit(x_train, y_train)
    y_pred = lasso.predict(x_test)
    mape = mean_absolute_percentage_error(y_test, y_pred)
    
    if mape < best_mape:
        best_mape = mape
        best_alpha = alpha

print(f"Best alpha: {best_alpha}")
print(f"Lowest MAPE: {best_mape}")

Best alpha: 0.001
Lowest MAPE: 4.4128318074269805e-05


1. The MAPE value decreases significantly to 4.4128318074115733e-05, when alpha = 0.001. 
2. With alpha set to 0.001, the regularization term is very weak. This means the model is almost unconstrained in terms of how large the coefficients can become. The model behaves much like an OLS model with only slight regularization.
3. An alpha as low as 0.001 might be appropriate if you have a large, noise-free dataset and believe that all features are likely to contribute meaningfully to the outcome.However, if the goal is to simplify the model or improve generalizability, a slightly higher alpha would usually be preferable to avoid overfitting and to keep the model more interpretable and robust.

In [ ]:
# Apply Ridge Regression to our data
from sklearn.linear_model import Ridge
from sklearn import linear_model as lm
model_ridge = lm.Ridge(alpha=1).fit(x_train,y_train)
model_ridge.coef_

array([ 1.29954993e+00,  1.69972328e+00,  6.07547327e+00,  2.09998671e+00,
       -9.01138066e-01, -2.76774698e-04,  1.39022392e-04, -1.29553546e-03,
        1.28574039e-04, -1.91998438e-04,  9.64691394e-04,  1.51252792e-04,
       -7.24514808e-03,  4.88963422e-05, -1.75560566e-03])

In [ ]:
# Predict on the test set using the fitted Ridge model
y_pred_ridge = model_ridge.predict(x_test)

In [ ]:
# Calculate the MAPE for the Ridge model
mape_ridge = mean_absolute_percentage_error(y_test, y_pred_ridge)
mape_ridge

np.float64(0.0013183099780278381)

1. The MAPE decreased significantly to 0.13%, much lower than the Lasso model when the MAPE is 4.43% where alpha was set at 1. 
2. Based on the comparison, we can conclude that Lasso is sensitive to the scale of alpha, and small changes in alpha can lead to significant differences in results. While Ridge provides more accurate predictions on this dataset when alpha is the same value with Lasso. This could be due to Ridge Regression’s ability to handle multicollinearity without setting coefficients to zero, effectively utilizing all features for a smoother fit. 

### Part D (25%)
1.	Use the demand data given in the table and develop an appropriate forecasting model (i.e., the tailored regularization discussed in the class—see your slides for more info) that exploits the available information given in the table as much as possible.
2.	Interpret your results.


In [ ]:
data = {
    'Month': list(range(1, 26)),
    'Demand': [100, 112, 107, 103, 91, 85, 84, 85, 79, 81, 134, 86, 99, 89, 111, 114, 118, 163, 193, 143, 144, 202, 158, 160, 144],
    'Advance_demand': [71, 30, 75, 64, 41, 51, 42, 51, 57, 49, 134, 52, 99, 56, 81, 79, 73, 163, 193, 99, 91, 202, 105, 101, 96]
}

df_d = pd.DataFrame(data)

In [ ]:
df_d

Month  Demand  Advance_demand
0       1     100              71
1       2     112              30
2       3     107              75
3       4     103              64
4       5      91              41
5       6      85              51
6       7      84              42
7       8      85              51
8       9      79              57
9      10      81              49
10     11     134             134
11     12      86              52
12     13      99              99
13     14      89              56
14     15     111              81
15     16     114              79
16     17     118              73
17     18     163             163
18     19     193             193
19     20     143              99
20     21     144              91
21     22     202             202
22     23     158             105
23     24     160             101
24     25     144              96

# GPT code

In [ ]:
from sklearn.metrics import mean_absolute_error

X = df_d[['Advance_demand']]  # 自变量 (特征)
y = df_d['Demand']  # 因变量 (目标)

# Step 2: 划分训练集和测试集
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

# Step 3: 运行 Lasso 模型
lasso = Lasso(alpha=1)  # alpha 是正则化强度的参数
lasso.fit(X_train, y_train)

# Step 4: 预测和评估模型
y_pred = lasso.predict(X_test)
mae = mean_absolute_error(y_test, y_pred)
print(f"Mean Absolute Error: {mae}")

# 打印模型系数
print(f"Model Coefficients: {lasso.coef_}")

Mean Absolute Error: 16.12293300907556
Model Coefficients: [0.69681898]


### create datasets and variables

In [ ]:
import pandas as pd
import numpy as np

# Create the dataset
data = {
    'Month': np.arange(1, 26),
    'Demand': [100, 112, 107, 103, 91, 85, 84, 85, 79, 81, 134, 86, 99, 89, 111, 114, 118, 163, 193, 143, 144, 202, 158, 160, 144],
    'Advance_demand': [71, 30, 75, 64, 41, 51, 42, 51, 57, 49, 134, 52, 99, 56, 81, 79, 73, 163, 193, 99, 91, 202, 105, 101, 96]
}

# Convert the data to a pandas DataFrame
df = pd.DataFrame(data)

# Prepare Y (target variable: Demand from month 2 to 25)
Y = df['Demand'][1:].values

# Prepare X (lagged variable: Demand from month 1 to 24)
X = df['Demand'][:-1].values

# Prepare L (advance demand from month 2 to 25 for constraints)
L = df['Advance_demand'][1:].values


### Build the Autoregressive Model (GLS) Predictor Matrix

In [ ]:
# Add a constant term (intercept) to build the predictor matrix for the autoregressive model
X_gls = np.vstack([np.ones_like(X), X]).T  # Convert to a 2D matrix


### Define the GLS Model Objective Function

In [ ]:
# Define the GLS model objective function (sum of squared residuals)
def objective_gls(beta, X, Y):
    residuals = Y - X.dot(beta)  # Calculate residuals
    return np.sum(residuals**2)  # Return the sum of squared residuals


 ### Define the Constraints

In [ ]:
# Define the constraint (predicted demand >= advance demand)
constraints = [{'type': 'ineq', 'fun': lambda beta: X_gls.dot(beta) - L}] 


### Solve the Optimization Problem

In [ ]:
from scipy.optimize import minimize

# Initial guess for beta (same length as the number of columns in X_gls)
beta_initial = np.zeros(X_gls.shape[1])

# Solve the constrained optimization problem for GLS
result_gls = minimize(objective_gls, beta_initial, args=(X_gls, Y), constraints=constraints)

# Extract the optimal coefficients
beta_gls_optimal = result_gls.x


### Predict Demand and Calculate the Error

In [ ]:
# Predict demand values using the optimal coefficients and ensure predictions are not lower than advance demand
Y_pred_gls = np.maximum(X_gls.dot(beta_gls_optimal), L)

# Calculate the Mean Absolute Percentage Error (MAPE)
mape_gls = np.mean(np.abs((Y - Y_pred_gls) / Y)) * 100

# Output the results
print(f"Optimal GLS coefficients: {beta_gls_optimal}")
print(f"GLS MAPE: {mape_gls:.2f}%")


Optimal GLS coefficients: [46.57142858  1.07936508]
GLS MAPE: 50.35%
